Fix the logic for CustomSchemeDetector to comply with the documentation.

olivandcode opened the pull request:
Modification of the CustomSchemeDetector logic to comly with https://developer.android.com/training/app-links/verify-applinks.

Commit:
Fix the logic for CustomSchemeDetector to comply with the documentation.

bjiang7 commented on checks/src/main/java/com/example/lint/checks/CustomSchemeDetector.kt:
```diff
@@ -31,68 +34,96 @@ import com.android.tools.lint.detector.api.XmlScanner
 import com.android.utils.childrenIterator
 import org.w3c.dom.Element
 
-/**
```
Nit: Could you write some more up-to-date KDoc then instead of just removing it?

olivandcode replied:
Done in the next commit

bjiang7 commented on checks/src/main/java/com/example/lint/checks/CustomSchemeDetector.kt:
```diff
@@ -31,68 +34,96 @@ import com.android.tools.lint.detector.api.XmlScanner
-                context.report(incident)
+    private fun hasViewAction(element: Element): Boolean {
+        val actions = element.getElementsByTagName(TAG_ACTION)
+        for (i in 0 until actions.length) {
```
Optional suggestion: you could use instead a higher-order function like filter. Or instead, just do a for each loop since you don't care about the actual index of the element, i.e. `for (action in actions) {...}`

https://kotlinlang.org/docs/coding-conventions.html#loops

Please fix throughout.

olivandcode replied:
Done in the next commit.

bjiang7 commented on checks/src/main/java/com/example/lint/checks/CustomSchemeDetector.kt:
```diff
@@ -31,68 +34,96 @@ import com.android.tools.lint.detector.api.XmlScanner
-/**
- * Detector flagging whether the application has issues verifying custom schemes (i.e. not http/https/file/ftp/ftps).
- */
 class CustomSchemeDetector : Detector(), XmlScanner {
```
Should we rename this class since it's only checking http/https intent filters not custom schemes?

olivandcode replied:
Indeed. I renamed it in the next commit.

bjiang7 commented on checks/src/main/java/com/example/lint/checks/CustomSchemeDetector.kt:
```diff
@@ -31,68 +34,96 @@ import com.android.tools.lint.detector.api.XmlScanner
+            return
+        }
+
+        for (i in 0 until dataElements.length) {
```
`for (element in dataElements) {...}`? Unless you plan on using the index for something, it's probably easier to just reference the elements directly in the for loop. Please fix throughout.

olivandcode replied:
Done in the next commit.

bjiang7 commented on checks/src/main/java/com/example/lint/checks/CustomSchemeDetector.kt:
```diff
@@ -31,68 +34,96 @@ import com.android.tools.lint.detector.api.XmlScanner
+        }
+
+        val dataElements = element.getElementsByTagName(TAG_DATA)
+        if (dataElements.length == 0) {
```
Do you even need this? Since you'll be iterating over dataElements, if dataElements is empty, it should skip any code in the for loop, right?

olivandcode replied:
Done in the next commit.

bjiang7 commented on checks/src/main/java/com/example/lint/checks/CustomSchemeDetector.kt:
```diff
@@ -31,68 +34,96 @@ import com.android.tools.lint.detector.api.XmlScanner
-                    val attribute = child.attributes.item(i)
-                    val name = attribute.localName ?: continue
-                    val value = attribute.nodeValue
+    private fun hasBrowsableCategory(element: Element): Boolean {
```
You could instead just have one function that takes in the category as a parameter, e.g. hasCategory(element: Element, category: String) and then reuse it for checking android.intent.category.BROWSABLE and android.intent.category.DEFAULT.

olivandcode replied:
Done in the next commit.

bjiang7 reviewed (changes requested):
Thanks for making this fix, Olivier!

Commit:
Refactor CustomSchemeDetector to MissingAutoVerifyDetector

- Renamed class to `MissingAutoVerifyDetector` to better reflect the specific check being performed.
- Added comprehensive KDoc.
- Refactored XML parsing to use idiomatic Kotlin `for-in` loops by adding a `NodeList.iterator()` extension function.
- Consolidated duplicate logic into a reusable `hasCategory()` helper function.
- Updated Issue ID and description to match the new class name.

bjiang7 commented:
Btw @olivandcode, I just remembered that you'll likely need to do 2 small additional changes:
1. Update the [README](https://github.com/google/android-security-lints/blob/main/README.md) to refer to this new class (since it's no longer called CustomSchemeDetector)
2. Update the sample app included in app/ to make sure it triggers the fixed AutoVerify lint check
